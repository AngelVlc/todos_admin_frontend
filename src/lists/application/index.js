export * from "./GetListByIdUseCase";
export * from "./DeleteListByIdUseCase";
export * from "./GetListsUseCase";
export * from "./CreateListUseCase";
export * from "./UpdateListUseCase";
export * from "./MoveListItemUseCase";
export * from "./IndexAllListsUseCase";
export * from "./GetSearchSecureKeyUseCase";
