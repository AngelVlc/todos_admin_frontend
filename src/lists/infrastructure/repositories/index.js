export * from "./ListsRepository";
export * from "./ListItemsRepository";
