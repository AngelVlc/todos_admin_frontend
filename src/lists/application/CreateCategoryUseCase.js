import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class CreateCategoryUseCase extends BaseUseCase {
  _categoriesRepository;

  constructor({ categoriesRepository }) {
    super();

    this._categoriesRepository = categoriesRepository;
  }

  static create() {
    const categoriesRepository = new CategoriesRepository();

    return new CreateCategoryUseCase({ categoriesRepository });
  }

  async execute(category) {
   return await this._categoriesRepository.create(category);
  }
}
