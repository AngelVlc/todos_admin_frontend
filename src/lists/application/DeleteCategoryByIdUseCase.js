import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class DeleteCategoryByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new DeleteCategoryByIdUseCase({ repository });
  }

  async execute(categoryId) {
    return await this._repository.deleteById(categoryId);
  }
}
