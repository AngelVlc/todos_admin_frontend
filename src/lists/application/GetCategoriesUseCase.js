import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { Category } from "../domain";
import { CategoriesRepository } from "../infrastructure/repositories";

export class GetCategoriesUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new GetCategoriesUseCase({ repository });
  }

  async execute() {
    const result = await this._repository.getAll();

    return result.map((item) => {
      return new Category(item);
    });
  }
}
