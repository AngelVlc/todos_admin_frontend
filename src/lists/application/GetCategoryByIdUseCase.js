import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { Category } from "../domain";
import { CategoriesRepository } from "../infrastructure/repositories";

export class GetCategoryByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new GetCategoryByIdUseCase({ repository });
  }

  async execute(id) {
    const result = await this._repository.getById(id);

    return new Category(result);
  }
}
