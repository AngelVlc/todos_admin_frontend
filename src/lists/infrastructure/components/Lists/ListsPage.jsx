import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetListsUseCase } from "../../../application/lists";

export const ListsPage = () => {
  const [lists, setLists] = useState([]);
  const { useCaseFactory } = useContext(AppContext);
  let history = useHistory();

  const getLists = useCallback(async () => {
    const getListsUseCase = useCaseFactory.get(GetListsUseCase);
    const data = await getListsUseCase.execute();

    setLists(data);
  }, [useCaseFactory]);

  useEffect(() => {
    getLists();
  }, [getLists]);

  const onNewClick = () => {
    history.push("/lists/new");
  };

  return (
    <div className="container">
      <h3 className="title">LISTS</h3>
      <nav className="breadcrumb" aria-label="breadcrumbs">
        <ul>
          <li>
            <Link to={`/`}>Home</Link>
          </li>
          <li className="is-active">
            <Link aria-current="page" to={`/lists`}>
              Lists
            </Link>
          </li>
        </ul>
      </nav>
      <table className="table">
        <thead>
          <tr>
            <td>Name</td>
            <td># Items</td>
            <td>
              <button
                className="button is-small"
                data-testid="addNew"
                onClick={() => onNewClick()}
              >
                <span className="icon is-small">
                  <i className="fas fa-plus"></i>
                </span>
              </button>
            </td>
          </tr>
        </thead>
        <tbody>
          {lists.length > 0 &&
            lists.map((list) => (
              <tr key={list.id}>
                <td>
                  <Link
                    className="has-text-black"
                    data-testid={`editList${list.id}`}
                    to={`/lists/${list.id}/edit`}
                  >
                    {list.name}
                  </Link>
                </td>
                <td>
                  <span>
                    <center>{list.itemsCount}</center>
                  </span>
                </td>
                <td>
                  <center>
                    <Link
                      className="has-text-black delete"
                      data-testid={`deleteList${list.id}`}
                      to={`/lists/${list.id}/delete`}
                    >
                    </Link>
                  </center>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};
