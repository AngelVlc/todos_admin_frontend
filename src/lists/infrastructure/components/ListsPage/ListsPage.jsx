import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link, useHistory } from "react-router-dom";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { GetListsUseCase } from "../../../application/lists";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";

export const ListsPage = () => {
  const [lists, setLists] = useState();
  const { useCaseFactory } = useContext(AppContext);
  let history = useHistory();

  const getLists = useCallback(async () => {
    const getListsUseCase = useCaseFactory.get(GetListsUseCase);
    const data = await getListsUseCase.execute();

    setLists(data);
  }, [useCaseFactory]);

  useEffect(() => {
    getLists();
  }, [getLists]);

  const onNewClick = () => {
    history.push("/lists/new");
  };

  return (
    <>
      {lists && (
        <div className="container">
          <Breadcrumb items={[{ url: "/lists", text: "Lists" }]} />
          <h3 className="title">LISTS</h3>
          <table className="table">
            <thead>
              <tr>
                <td>Name</td>
                <td># Items</td>
                <td>
                  <button
                    className="button is-small"
                    data-testid="addNew"
                    onClick={() => onNewClick()}
                  >
                    <span className="icon is-small">
                      <i className="fas fa-plus"></i>
                    </span>
                  </button>
                </td>
              </tr>
            </thead>
            <tbody>
              {lists.length > 0 &&
                lists.map((list) => (
                  <tr key={list.id}>
                    <td>
                      <Link
                        className="has-text-black"
                        data-testid={`viewList${list.id}`}
                        to={`/lists/${list.id}`}
                      >
                        {list.name}
                      </Link>
                    </td>
                    <td>
                      <span>
                        <center>{list.itemsCount}</center>
                      </span>
                    </td>
                    <td>
                      <center>
                        <Link
                          className="has-text-black delete"
                          data-testid={`deleteList${list.id}`}
                          to={`/lists/${list.id}/delete`}
                        ></Link>
                      </center>
                    </td>
                  </tr>
                ))}
            </tbody>
          </table>
        </div>
      )}
    </>
  );
};
