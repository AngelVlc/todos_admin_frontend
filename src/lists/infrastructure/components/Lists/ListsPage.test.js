import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { ListsPage } from "./ListsPage";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { List } from "../../../domain";
import { GetListsUseCase, GetSearchSecureKeyUseCase, GetCategoriesUseCase } from "../../../application";

afterEach(cleanup);

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const history = createMemoryHistory();

const renderWithContextAndRouter = () => {
  const fakeGetListsUseCase = {
    execute: () => [
      new List({ id: 1, name: "list1", categoryId: 1, itemsCount: 4 }),
      new List({ id: 2, name: "list2", categoryId: null, itemsCount: 6 }),
    ],
  };

  const fakeGetSearchSecureKeyUseCase = {
    execute: () => "",
  };

  const fakeGetCategoriesUseCase = {
    execute: () => [
      { id: 1, name: "category1" },
      { id: 2, name: "category2" },
    ],
  };

  const useCaseFactory = {
    get: (useCase) => {
      switch(useCase) {
        case GetListsUseCase:
          return fakeGetListsUseCase;
        case GetSearchSecureKeyUseCase:
          return fakeGetSearchSecureKeyUseCase;
        case GetCategoriesUseCase:
          return fakeGetCategoriesUseCase;
      };
    },
  };

  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <Router history={history}>
          <ListsPage />
        </Router>
      </AppContext.Provider>
    ),
  };
};

describe("ListsPage", () => {
  it("should match the snapshot", async () => {
    let fragment;
    await act(async () => {
      const { asFragment } = renderWithContextAndRouter();
      fragment = asFragment;
    });

    expect(fragment()).toMatchSnapshot();
  });

  it("should add a new list", async () => {
    const container = renderWithContextAndRouter();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("addNew"));
    });

    expect(mockHistoryPush).toHaveBeenCalled();
    expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/new");
    mockHistoryPush.mockClear();
  });

  it("should view the list", async () => {
    const container = renderWithContextAndRouter();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("viewList2"));
    });

    expect(mockHistoryPush).toHaveBeenCalled();
    expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/2");
    mockHistoryPush.mockClear();
  });

  it("should delete the list", async () => {
    const container = renderWithContextAndRouter();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("deleteList2"));
    });

    expect(history.location.pathname).toBe("/lists/2/delete");
  });
});
