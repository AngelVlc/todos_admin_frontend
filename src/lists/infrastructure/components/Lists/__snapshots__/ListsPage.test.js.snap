// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListsPage should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal"
    data-testid="modal"
  >
    <div
      class="modal-background"
    />
    <div
      class="modal-content"
    >
      <div
        class="box"
      >
        <div
          class="content"
        >
          <center>
            <h5>
              Algolia not configured
            </h5>
          </center>
        </div>
        <div
          class="field is-grouped is-justify-content-center"
        >
          <div
            class="control"
          >
            <button
              class="button"
              data-testid="modalCancel"
            >
              CANCEL
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      aria-label="close"
      class="modal-close is-large"
      data-testid="modalClose"
    />
  </div>
  <div
    class="container"
  >
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/lists"
          >
            Lists
          </a>
        </li>
      </ul>
    </nav>
    <h3
      class="title"
    >
      LISTS
    </h3>
    <table
      class="table"
    >
      <thead>
        <tr>
          <td>
            Name
          </td>
          <td>
            Category
          </td>
          <td>
            # Items
          </td>
          <td>
            <button
              class="button is-small mr-2"
              data-testid="search"
            >
              <span
                class="icon is-small"
              >
                <i
                  class="fas fa-search"
                />
              </span>
            </button>
            <button
              class="button is-small"
              data-testid="addNew"
            >
              <span
                class="icon is-small"
              >
                <i
                  class="fas fa-plus"
                />
              </span>
            </button>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr
          class="is-clickable"
          data-testid="viewList1"
        >
          <td>
            list1
          </td>
          <td>
            category1
          </td>
          <td>
            <center>
              4
            </center>
          </td>
          <td>
            <center>
              <a
                class="has-text-black delete"
                data-testid="deleteList1"
                href="/lists/1/delete"
              />
            </center>
          </td>
        </tr>
        <tr
          class="is-clickable"
          data-testid="viewList2"
        >
          <td>
            list2
          </td>
          <td />
          <td>
            <center>
              6
            </center>
          </td>
          <td>
            <center>
              <a
                class="has-text-black delete"
                data-testid="deleteList2"
                href="/lists/2/delete"
              />
            </center>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
